### ISSUE TITLE
Possible edge case in channel manager

### ISSUE THREAD

**Issue by yigit** [dropbox/Store#41]
There might be an edge case in channel manager where it might idle forever.
have not tried but we should check: (buffer = 0)
* add a downstream that is suspending on collect
* dispatch value from upstream
* add another downstream
* remove first downstream

I believe it will be broken such that the first value will never be acknowledged but the second downstream will just await forever.

Seems like we need some more tracking on message acks

**Comment by digitalbuddha** [comment]
Double checking - you mean the first collector is suspended and never acks?  

One solution would be to flush/ack all when the last collector is removed.

**Comment by yigit** [comment]
yea, got a test failing:
https://github.com/yigit/Core/commit/237f84d235ac29f90c9d4e2c2fe819c9d211d813

We need to make a decision here. If we give that value to the second downstream, we'll violate the rule of "happens after" when buffer is 0.
One option is to ack messages when a downstream is closed and that message was not sent to any other upstream.
Another option, which i'm leaning towards is to immediately unblock upstream when a new downstream arrives if buffer is 0. I like that option better because the new downstream will be blocked by the other one if we don't do this, which might cause very weird use cases.

tl;dr; I'm try to make the rule such that 
* if buffer is 0
* and a new downstream comes
* immediately unblock upstream

This way, a later arrival is not blocked for messages that it will never receive and a closing channel w/o acking does not cause any issues.

wdyt?

**Comment by eyalgu** [comment]
This rule may violate the "only one active fetch" rule though.
If we want to preserve both rules we would probably need to be more granular here:

have `ChannelManager` track the state of the upstream to distigush between `active` and `awaiting ack` and only kick the upstream if it's in awaiting ack

If we're already tracking `awaiting ack` we can also track which channel(s) we're awaiting on and unblock the upstream only if we've removed the last channel that could have acked that upstream (and there are still more downstreams). 
This extra bit of logic will prevent us from starting the next fetch while some channels may still be processing the previous one. This situation will create a new edge case where when buffer=0 a downstream could will process an old value even after a new one has been fetched by the unblocked uptream.

**Comment by yigit** [comment]
Not sure if I understand correctly, mind creating a test case for it?
Last channel thing does not fix the issue i'm trying to address.
Right now, if a new downstream arrives while the upstream is blocked on ack, the new downstream waits for a value ack that it would've never received, practically suspending it due to another collector being slow.

> This extra bit of logic will prevent us from starting the next fetch while some channels may still be processing the previous one. This situation will create a new edge case where when buffer=0 a downstream could receive an old value even after a new one has been fetched by the unblocked uptream.
We actually want that behavior. The blocked downstreams will just buffer the new value in their channel until they can consume. It is OK for them to receive the old value as it was produced after they've subscribed,they get those values right now anyways  (w/ their unlimited buffer). if another downstream is faster If they want to always get latest, they can conflate.

The goal of the `ack`ing is not ensuring that every upstream value is processed before moving to the next one. The goal of it is to not produce values when there are no downstream to receive a value.
When we get a new downstream, it means we have one that can receive a value, hence should unblock upstream.

Does that make sense ?

**Comment by eyalgu** [comment]
failing test due to new edge case: c63ac7553290e5d6c6f6ba056699ae21191ed619

**Comment by yigit** [comment]
that is not a failure, it is the expected case.
We don't conflate values.
The upstream is "a", "b", "c"
If c1 takes 2 values, it is expected to be "a", "b". If it waits for the 3rd one, it will get "c".

**Comment by eyalgu** [comment]
yeah come to think of it. something similar will happen today when we have more than 1 channel and one of the collectors is slow
